Erro ao acessar o Badger no inicio da segunda etapa do transform

Estou com o seguinte erro durante o transform :
```
docker compose run --rm minha-receita transform -d /mnt/data // ou com /mnt/data/sample o erro persistia

Error: could not create badger storage: error creating badger key-value object: Cannot acquire directory lock on "/tmp/minha-receita-badger--202410280015481995141590".  Another process is using this Badger database. error: resource temporarily unavailable
Usage:
  minha-receita transform [flags]

Flags:
  [...]

could not create badger storage: error test creating badger key-value object: Cannot acquire directory lock on "/tmp/minha-receita-badger--202410280015481995141590".  Another process is using this Badger database. error: resource temporarily unavailable
```

Depois de 2 dias pesquisando e entendendo um pouco melhor de como o transform funciona mesmo sabendo quase nada de GO identifiquei que o erro acontecia ao iniciar a etapa 2 conforme descrito no [arquivo CONTRIBUTING](https://github.com/cuducos/minha-receita/blob/main/CONTRIBUTING.md#estrat%C3%A9gia-de-carregamento-dos-dados-no-postgresql).

Pelo que identifiquei quando o StepTwo ao tentar abrir a conexão com o badger ele não consegue pois o StepOne não fecha a conexão ao finalizar.

`db, err := badger.Open(opt.WithLogger(&badgerLogger{}))`

Então fechando a conexão ao final do StepOne (apenas copiando o comportamento do StepTwo) eu consegui rodar com os arquivos de sample sem o erro acontecer novamente e olhando o banco a única coisa que percebi de estranho foi que em todos o campo de "qsa" estava vazio, mas acredito que isso possa ser uma questão do arquivo de sample, acabei de colocar pra executar com os arquivos reais e assim que terminar também trago um feedback se necessário. Também queria saber qual o comportamento esperado do StepOne pra ele não fechar a conexão anteriormente e se eu ter adicionado esse comportamento pode ter algum impacto nos dados transformados.

```
func runStepOne(dir string, l lookups, isolated bool) (string, error) {
	
	tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
	
	if err != nil {
		return "", fmt.Errorf("error creating temporary key-value storage: %w", err)
	}
	
	kv, err := newBadgerStorage(tmp)
	
	if err != nil {
		return "", fmt.Errorf("could not create badger storage: %w", err)
	}
	
	if err := kv.load(dir, &l); err != nil {
		return "", fmt.Errorf("error loading data to badger: %w", err)
	}
	
	if isolated {
		fmt.Println(kv.path)
	}
	
	if !isolated {
		defer kv.close()
	}
	
	return kv.path, nil
}
```

Dados da máquina:
VM com ubuntu-server 24.04
Ram 16gb
Armazenamento 250gb
Executei o projeto apenas usando Docker, mas também tinha testado usando GO 1.23 antes de testar a questão de fechar a conexão mas tive o mesmo erro e voltei a usar o Docker.

Tentei aqui mas ao buildar com Docker Compose tive esse erro:

```
 => ERROR [minha-receita build 7/7] RUN go build -o /usr/bin/minha-receita                                                                                           86.6s
------
 > [minha-receita build 7/7] RUN go build -o /usr/bin/minha-receita:
85.79 # github.com/cuducos/minha-receita/transform
85.79 transform/transform.go:103:47: cannot use kv (variable of type *badgerStorage) as kvStorage value in argument to createJSONRecordsTask: *badgerStorage does not implement kvStorage (wrong type for method close)
85.79           have close(bool) error
85.79           want close() error
------
failed to solve: process "/bin/sh -c go build -o /usr/bin/minha-receita" did not complete successfully: exit code: 1
```

@cuducos eu tentei rodar a nova branch, mas tive esse problema: 

```
2024/10/31 10:15:09 Loading Estabelecimentos files…
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x1014b3430]

goroutine 88710630 [running]:
github.com/cuducos/minha-receita/transform.(*source).countLinesFor(0x14044972a80, 0x0, 0x14065bcc0e0, 0x14065bcc150)
	/Users/projects/minha-receita/transform/source.go:90 +0x60
created by github.com/cuducos/minha-receita/transform.(*source).countLines in goroutine 1
	/Users/projects/minha-receita/transform/source.go:112 +0x88
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x1014b3430]

goroutine 88710625 [running]:
github.com/cuducos/minha-receita/transform.(*source).countLinesFor(0x14044972a80, 0x0, 0x14065bcc0e0, 0x14065bcc150)
	/Users/projects/minha-receita/transform/source.go:90 +0x60
created by github.com/cuducos/minha-receita/transform.(*source).countLines in goroutine 1
	/Users/projects/minha-receita/transform/source.go:112 +0x88
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x1014b3430]

goroutine 88710629 [running]:
github.com/cuducos/minha-receita/transform.(*source).countLinesFor(0x14044972a80, 0x0, 0x14065bcc0e0, 0x14065bcc150)
	/Users/projects/minha-receita/transform/source.go:90 +0x60
created by github.com/cuducos/minha-receita/transform.(*source).countLines in goroutine 1
	/Users/projects/minha-receita/transform/source.go:112 +0x88
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x1014b3430]

goroutine 88710628 [running]:
github.com/cuducos/minha-receita/transform.(*source).countLinesFor(0x14044972a80, 0x0, 0x14065bcc0e0, 0x14065bcc150)
	/Users/projects/minha-receita/transform/source.go:90 +0x60
created by github.com/cuducos/minha-receita/transform.(*source).countLines in goroutine 1
	/Users/projects/minha-receita/transform/source.go:112 +0x88
```

--- a/go.mod
+++ b/go.mod
@@ -49,5 +49,3 @@ require (
 	google.golang.org/grpc v1.65.0 // indirect
 	google.golang.org/protobuf v1.34.2 // indirect
 )
-
-// +heroku goVersion go1.22

--- a/transform/company_test.go
+++ b/transform/company_test.go
@@ -108,7 +108,7 @@ func TestNewCompany(t *testing.T) {
 	}
 
 	t.Run("with privacy", func(t *testing.T) {
-		tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
+		tmp, err := os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
 		if err != nil {
 			t.Fatal("error creating temporary key-value storage: %w", err)
 		}
@@ -117,7 +117,7 @@ func TestNewCompany(t *testing.T) {
 		if err != nil {
 			t.Errorf("expected no error creating badger, got %s", err)
 		}
-		defer kv.close(false)
+		defer kv.close()
 		lookups, err := newLookups(testdata)
 		if err != nil {
 			t.Errorf("expected no errors creating look up tables, got %v", err)
@@ -264,7 +264,7 @@ func TestNewCompany(t *testing.T) {
 		}
 	})
 	t.Run("without privacy", func(t *testing.T) {
-		tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
+		tmp, err := os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
 		if err != nil {
 			t.Fatal("error creating temporary key-value storage: %w", err)
 		}
@@ -273,7 +273,7 @@ func TestNewCompany(t *testing.T) {
 		if err != nil {
 			t.Errorf("expected no error creating badger, got %s", err)
 		}
-		defer kv.close(false)
+		defer kv.close()
 		lookups, err := newLookups(testdata)
 		if err != nil {
 			t.Errorf("expected no errors creating look up tables, got %v", err)

--- a/transform/company_test.go
+++ b/transform/company_test.go
@@ -117,7 +117,7 @@ func TestNewCompany(t *testing.T) {
 		if err != nil {
 			t.Errorf("expected no error creating badger, got %s", err)
 		}
-		defer kv.close()
+		defer kv.close(false)
 		lookups, err := newLookups(testdata)
 		if err != nil {
 			t.Errorf("expected no errors creating look up tables, got %v", err)
@@ -273,7 +273,7 @@ func TestNewCompany(t *testing.T) {
 		if err != nil {
 			t.Errorf("expected no error creating badger, got %s", err)
 		}
-		defer kv.close()
+		defer kv.close(false)
 		lookups, err := newLookups(testdata)
 		if err != nil {
 			t.Errorf("expected no errors creating look up tables, got %v", err)

--- a/transform/kv.go
+++ b/transform/kv.go
@@ -162,31 +162,25 @@ func (kv *badgerStorage) enrichCompany(c *company) error {
 	return nil
 }
 
-func (b *badgerStorage) close(k bool) error {
-	b.db.Close()
-	if !k {
-		if err := os.RemoveAll(b.path); err != nil {
-			return fmt.Errorf("error cleaning up badger storage directory: %w", err)
-		}
-	}
-	return nil
+func (b *badgerStorage) close() error {
+	return b.db.Close()
 }
 
-type badgerLogger struct{}
+type noLogger struct{}
 
-func (*badgerLogger) Errorf(string, ...interface{})   {}
-func (*badgerLogger) Warningf(string, ...interface{}) {}
-func (*badgerLogger) Infof(string, ...interface{})    {}
-func (*badgerLogger) Debugf(string, ...interface{})   {}
+func (*noLogger) Errorf(string, ...interface{})   {}
+func (*noLogger) Warningf(string, ...interface{}) {}
+func (*noLogger) Infof(string, ...interface{})    {}
+func (*noLogger) Debugf(string, ...interface{})   {}
 
 func newBadgerStorage(dir string) (*badgerStorage, error) {
-	var err error
-	var opt badger.Options
+	opt := badger.DefaultOptions(dir)
 	if os.Getenv("DEBUG") != "" {
 		log.Output(1, fmt.Sprintf("Creating temporary key-value storage at %s", dir))
+	} else {
+		opt = opt.WithLogger(&noLogger{})
 	}
-	opt = badger.DefaultOptions(dir)
-	db, err := badger.Open(opt.WithLogger(&badgerLogger{}))
+	db, err := badger.Open(opt)
 	if err != nil {
 		return nil, fmt.Errorf("error creating badger key-value object: %w", err)
 	}

--- a/transform/kv.go
+++ b/transform/kv.go
@@ -162,9 +162,9 @@ func (kv *badgerStorage) enrichCompany(c *company) error {
 	return nil
 }
 
-func (b *badgerStorage) close() error {
+func (b *badgerStorage) close(k bool) error {
 	b.db.Close()
-	if b.path != "" {
+	if !k && b.path != ""  {
 		if err := os.RemoveAll(b.path); err != nil {
 			return fmt.Errorf("error cleaning up badger storage directory: %w", err)
 		}

--- a/transform/kv.go
+++ b/transform/kv.go
@@ -164,7 +164,7 @@ func (kv *badgerStorage) enrichCompany(c *company) error {
 
 func (b *badgerStorage) close(k bool) error {
 	b.db.Close()
-	if !k && b.path != ""  {
+	if !k {
 		if err := os.RemoveAll(b.path); err != nil {
 			return fmt.Errorf("error cleaning up badger storage directory: %w", err)
 		}

--- a/transform/kv_test.go
+++ b/transform/kv_test.go
@@ -19,7 +19,7 @@ func TestBadgerStorageClose(t *testing.T) {
 	if err != nil {
 		t.Errorf("expected no error creating badger storage, got %s", err)
 	}
-	if err := kv.close(); err != nil {
+	if err := kv.close(false); err != nil {
 		t.Errorf("expected no error closing badger storage, got %s", err)
 	}
 	if _, err := os.Stat(kv.path); err == nil || !os.IsNotExist(err) {
@@ -74,7 +74,7 @@ func TestLoad(t *testing.T) {
 	if err != nil {
 		t.Fatalf("could not create badger storage: %s", err)
 	}
-	defer kv.close()
+	defer kv.close(false)
 	if err := kv.load(testdata, &l); err != nil {
 		t.Errorf("expected no error loading data, got %s", err)
 	}
@@ -103,7 +103,7 @@ func TestEnrichCompany(t *testing.T) {
 	if err != nil {
 		t.Fatalf("could not create badger storage: %s", err)
 	}
-	defer kv.close()
+	defer kv.close(false)
 	if err := kv.load(testdata, &l); err != nil {
 		t.Errorf("expected no error loading data, got %s", err)
 	}

--- a/transform/kv_test.go
+++ b/transform/kv_test.go
@@ -10,7 +10,7 @@ import (
 )
 
 func TestBadgerStorageClose(t *testing.T) {
-	tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
+	tmp, err := os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
 	if err != nil {
 		t.Fatal("error creating temporary key-value storage: %w", err)
 	}
@@ -19,11 +19,8 @@ func TestBadgerStorageClose(t *testing.T) {
 	if err != nil {
 		t.Errorf("expected no error creating badger storage, got %s", err)
 	}
-	if err := kv.close(false); err != nil {
+	if err := kv.close(); err != nil {
 		t.Errorf("expected no error closing badger storage, got %s", err)
-	}
-	if _, err := os.Stat(kv.path); err == nil || !os.IsNotExist(err) {
-		t.Errorf("expected %s to be gone, but got %s when opening it", kv.path, err)
 	}
 }
 
@@ -65,7 +62,7 @@ func TestLoad(t *testing.T) {
 	if err != nil {
 		t.Fatalf("could not create lookups: %s", err)
 	}
-	tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
+	tmp, err := os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
 	if err != nil {
 		t.Fatal("error creating temporary key-value storage: %w", err)
 	}
@@ -74,7 +71,7 @@ func TestLoad(t *testing.T) {
 	if err != nil {
 		t.Fatalf("could not create badger storage: %s", err)
 	}
-	defer kv.close(false)
+	defer kv.close()
 	if err := kv.load(testdata, &l); err != nil {
 		t.Errorf("expected no error loading data, got %s", err)
 	}
@@ -94,7 +91,7 @@ func TestEnrichCompany(t *testing.T) {
 	if err != nil {
 		t.Fatalf("could not create lookups: %s", err)
 	}
-	tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
+	tmp, err := os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
 	if err != nil {
 		t.Fatal("error creating temporary key-value storage: %w", err)
 	}
@@ -103,7 +100,7 @@ func TestEnrichCompany(t *testing.T) {
 	if err != nil {
 		t.Fatalf("could not create badger storage: %s", err)
 	}
-	defer kv.close(false)
+	defer kv.close()
 	if err := kv.load(testdata, &l); err != nil {
 		t.Errorf("expected no error loading data, got %s", err)
 	}

--- a/transform/transform.go
+++ b/transform/transform.go
@@ -19,8 +19,6 @@ const (
 	// BatchSize determines the size of the batches used to create the initial JSON
 	// data in the database.
 	BatchSize = 8192
-
-	badgerFilePrefix = "minha-receita-badger-"
 )
 
 type database interface {
@@ -33,7 +31,7 @@ type database interface {
 type kvStorage interface {
 	load(string, *lookups) error
 	enrichCompany(*company) error
-	close(bool) error
+	close() error
 }
 
 type mode int
@@ -75,31 +73,24 @@ func saveUpdatedAt(db database, dir string) error {
 	return db.MetaSave("updated-at", string(v))
 }
 
-func runStepOne(dir string, l lookups, isolated bool) (string, error) {
-	tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
-	if err != nil {
-		return "", fmt.Errorf("error creating temporary key-value storage: %w", err)
-	}
-	kv, err := newBadgerStorage(tmp)
-	if err != nil {
-		return "", fmt.Errorf("could not create badger storage: %w", err)
-	}
-	defer kv.close(isolated)
-	if err := kv.load(dir, &l); err != nil {
-		return "", fmt.Errorf("error loading data to badger: %w", err)
-	}
-	if isolated {
-		fmt.Println(kv.path)
-	}
-	return kv.path, nil
-}
-
-func runStepTwo(dir string, tmp string, db database, l lookups, maxParallelDBQueries, batchSize int, privacy, isolated bool) error {
-	kv, err := newBadgerStorage(tmp)
+func runStepOne(dir string, pth string, l lookups) error {
+	kv, err := newBadgerStorage(pth)
 	if err != nil {
 		return fmt.Errorf("could not create badger storage: %w", err)
 	}
-	defer kv.close(isolated)
+	defer kv.close()
+	if err := kv.load(dir, &l); err != nil {
+		return fmt.Errorf("error loading data to badger: %w", err)
+	}
+	return nil
+}
+
+func runStepTwo(dir string, pth string, db database, l lookups, maxParallelDBQueries, batchSize int, privacy bool) error {
+	kv, err := newBadgerStorage(pth)
+	if err != nil {
+		return fmt.Errorf("could not create badger storage: %w", err)
+	}
+	defer kv.close()
 	j, err := createJSONRecordsTask(dir, db, &l, kv, batchSize, privacy)
 	if err != nil {
 		return fmt.Errorf("error creating new task for venues in %s: %w", dir, err)
@@ -112,27 +103,38 @@ func runStepTwo(dir string, tmp string, db database, l lookups, maxParallelDBQue
 
 // Transform the downloaded files for company venues creating a database record
 // per CNPJ
-func Transform(dir string, db database, maxParallelDBQueries, batchSize int, privacy, s1 bool, s2 string) error {
+func Transform(dir string, db database, max, s int, p, s1 bool, s2 string) error {
 	m, err := transformMode(s1, s2)
 	if err != nil {
 		return fmt.Errorf("error determining transform mode: %w", err)
 	}
-	var tmp string
+	var pth string
+	if m == stepTwo {
+		pth = s2
+	} else {
+		pth, err = os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
+	}
+	if err != nil {
+		return fmt.Errorf("error creating temporary key-value storage: %w", err)
+	}
+	defer os.RemoveAll(pth)
 	l, err := newLookups(dir)
 	if err != nil {
 		return fmt.Errorf("error creating look up tables from %s: %w", dir, err)
 	}
-	if m != stepTwo {
-		tmp, err = runStepOne(dir, l, m == stepOne)
-		if err != nil {
-			return fmt.Errorf("error creating key-value storage: %w", err)
+	switch m {
+	case stepOne:
+		if err := runStepOne(dir, pth, l); err != nil {
+			return err
 		}
-	}
-	if m != stepOne {
-		if s2 != "" {
-			tmp = s2
+		fmt.Println(pth)
+	case stepTwo:
+		return runStepTwo(dir, pth, db, l, max, s, p)
+	case both:
+		if err := runStepOne(dir, pth, l); err != nil {
+			return err
 		}
-		return runStepTwo(dir, tmp, db, l, maxParallelDBQueries, batchSize, privacy, m == stepTwo)
+		return runStepTwo(dir, pth, db, l, max, s, p)
 	}
 	return nil
 }

--- a/transform/transform.go
+++ b/transform/transform.go
@@ -33,7 +33,7 @@ type database interface {
 type kvStorage interface {
 	load(string, *lookups) error
 	enrichCompany(*company) error
-	close() error
+	close(bool) error
 }
 
 type mode int

--- a/transform/transform.go
+++ b/transform/transform.go
@@ -84,6 +84,7 @@ func runStepOne(dir string, l lookups, isolated bool) (string, error) {
 	if err != nil {
 		return "", fmt.Errorf("could not create badger storage: %w", err)
 	}
+	defer kv.close(isolated)
 	if err := kv.load(dir, &l); err != nil {
 		return "", fmt.Errorf("error loading data to badger: %w", err)
 	}
@@ -98,9 +99,7 @@ func runStepTwo(dir string, tmp string, db database, l lookups, maxParallelDBQue
 	if err != nil {
 		return fmt.Errorf("could not create badger storage: %w", err)
 	}
-	if !isolated {
-		defer kv.close()
-	}
+	defer kv.close(isolated)
 	j, err := createJSONRecordsTask(dir, db, &l, kv, batchSize, privacy)
 	if err != nil {
 		return fmt.Errorf("error creating new task for venues in %s: %w", dir, err)

--- a/transform/venues_test.go
+++ b/transform/venues_test.go
@@ -18,7 +18,7 @@ func TestTaskRun(t *testing.T) {
 	if err != nil {
 		t.Errorf("expected no error creating badger, got %s", err)
 	}
-	defer kv.close()
+	defer kv.close(false)
 	lookups, err := newLookups(testdata)
 	if err != nil {
 		t.Errorf("expected no errors creating look up tables, got %v", err)

--- a/transform/venues_test.go
+++ b/transform/venues_test.go
@@ -9,7 +9,7 @@ import (
 
 func TestTaskRun(t *testing.T) {
 	db := newTestDB(t)
-	tmp, err := os.MkdirTemp("", fmt.Sprintf("%s-%s", badgerFilePrefix, time.Now().Format("20060102150405")))
+	tmp, err := os.MkdirTemp("", fmt.Sprintf("minha-receita-%s-*", time.Now().Format("20060102150405")))
 	if err != nil {
 		t.Fatal("error creating temporary key-value storage: %w", err)
 	}
@@ -18,7 +18,7 @@ func TestTaskRun(t *testing.T) {
 	if err != nil {
 		t.Errorf("expected no error creating badger, got %s", err)
 	}
-	defer kv.close(false)
+	defer kv.close()
 	lookups, err := newLookups(testdata)
 	if err != nil {
 		t.Errorf("expected no errors creating look up tables, got %v", err)

